File: backend/app/core/fiscal.py
from datetime import date

# InfoBeans' fiscal year — assumed April-to-March (the common Indian FY convention),
# labeled by the calendar year it ends in (Apr 2025-Mar 2026 = FY2026). This is an
# open item to confirm with stakeholders (see docs/plan); change this one constant
# if it's wrong, everything downstream derives from it.
FISCAL_YEAR_START_MONTH = 4


def fiscal_year_of(d: date) -> int:
    return d.year + 1 if d.month >= FISCAL_YEAR_START_MONTH else d.year


def fiscal_quarter_of(d: date) -> int:
    months_since_start = (d.month - FISCAL_YEAR_START_MONTH) % 12
    return months_since_start // 3 + 1


def fiscal_month_of(d: date) -> int:
    """1-12 with the fiscal year's first month (April) as 1 — for ordering a
    monthly trend chart Apr→Mar instead of calendar Jan→Dec."""
    return (d.month - FISCAL_YEAR_START_MONTH) % 12 + 1


def current_fiscal_year() -> int:
    return fiscal_year_of(date.today())

File: frontend/src/lib/fiscal.ts
// Mirrors backend/app/core/fiscal.py — keep the two in sync if this ever changes.
// InfoBeans' fiscal year is assumed April-to-March (the common Indian FY convention),
// labeled by the calendar year it ends in (Apr 2025-Mar 2026 = FY2026). Unconfirmed
// with stakeholders — change FISCAL_YEAR_START_MONTH if it's wrong.
export const FISCAL_YEAR_START_MONTH = 4

export function fiscalYearOf(d: Date): number {
  const month = d.getMonth() + 1 // JS months are 0-indexed
  return month >= FISCAL_YEAR_START_MONTH ? d.getFullYear() + 1 : d.getFullYear()
}

export function currentFiscalYear(): number {
  return fiscalYearOf(new Date())
}

export const EARLIEST_FISCAL_YEAR = 2026

/** Selectable fiscal years, most recent first. */
export function fiscalYearOptions(): number[] {
  const years: number[] = []
  for (let y = currentFiscalYear(); y >= EARLIEST_FISCAL_YEAR; y--) years.push(y)
  return years
}

/** "FY27" rather than "FY2027" — the short form used everywhere in the UI. */
export function fyLabel(year: number): string {
  return `FY${String(year).slice(-2)}`
}
